flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/DefaultJobLeaderService.java

```diff
@@ -0,0 +1,451 @@
+	private JobLeaderListener jobLeaderListener;
+
+	public DefaultJobLeaderService(
+			UnresolvedTaskManagerLocation location,
+			RetryingRegistrationConfiguration retryingRegistrationConfiguration) {
+		this.ownLocation = Preconditions.checkNotNull(location);
+		this.retryingRegistrationConfiguration = Preconditions.checkNotNull(retryingRegistrationConfiguration);
+
+		// Has to be a concurrent hash map because tests might access this service
+		// concurrently via containsJob
+		jobLeaderServices = new ConcurrentHashMap<>(4);
+
+		state = DefaultJobLeaderService.State.CREATED;
+
+		ownerAddress = null;
+		rpcService = null;
+		highAvailabilityServices = null;
+		jobLeaderListener = null;
+	}
+
+	// -------------------------------------------------------------------------------
+	// Methods
+	// -------------------------------------------------------------------------------
+
+	@Override
+	public void start(
+		final String initialOwnerAddress,
+		final RpcService initialRpcService,
+		final HighAvailabilityServices initialHighAvailabilityServices,
+		final JobLeaderListener initialJobLeaderListener) {
+
+		if (DefaultJobLeaderService.State.CREATED != state) {
+			throw new IllegalStateException("The service has already been started.");
+		} else {
+			LOG.info("Start job leader service.");
+
+			this.ownerAddress = Preconditions.checkNotNull(initialOwnerAddress);
+			this.rpcService = Preconditions.checkNotNull(initialRpcService);
+			this.highAvailabilityServices = Preconditions.checkNotNull(initialHighAvailabilityServices);
+			this.jobLeaderListener = Preconditions.checkNotNull(initialJobLeaderListener);
+			state = DefaultJobLeaderService.State.STARTED;
+		}
+	}
+
+	@Override
+	public void stop() throws Exception {
+		LOG.info("Stop job leader service.");
+
+		if (DefaultJobLeaderService.State.STARTED == state) {
+
+			for (Tuple2<LeaderRetrievalService, DefaultJobLeaderService.JobManagerLeaderListener> leaderRetrievalServiceEntry: jobLeaderServices.values()) {
+				LeaderRetrievalService leaderRetrievalService = leaderRetrievalServiceEntry.f0;
+				DefaultJobLeaderService.JobManagerLeaderListener jobManagerLeaderListener = leaderRetrievalServiceEntry.f1;
+
+				jobManagerLeaderListener.stop();
+				leaderRetrievalService.stop();
+			}
+
+			jobLeaderServices.clear();
+		}
+
+		state = DefaultJobLeaderService.State.STOPPED;
+	}
+
+	@Override
+	public void removeJob(JobID jobId) throws Exception {
+		Preconditions.checkState(DefaultJobLeaderService.State.STARTED == state, "The service is currently not running.");
+
+		Tuple2<LeaderRetrievalService, DefaultJobLeaderService.JobManagerLeaderListener> entry = jobLeaderServices.remove(jobId);
+
+		if (entry != null) {
+			LOG.info("Remove job {} from job leader monitoring.", jobId);
+
+			LeaderRetrievalService leaderRetrievalService = entry.f0;
+			DefaultJobLeaderService.JobManagerLeaderListener jobManagerLeaderListener = entry.f1;
+
+			leaderRetrievalService.stop();
```

Comment:
If this throws an exception, where is the leader listener being closed?

Reply:
I think it is not properly closed. I will add a fix for it.

Reply:
I've added another hotfix commit which moves the exception handling into the `DefaultJobLeaderService`. I think this is ok because we stop the `JobManagerLeaderListener` so that we will no longer react to leader change notifications.